local-server/src/lib.rs

```diff
@@ -224,15 +227,44 @@ async fn linkup_request_handler(
 
     let extra_headers = get_additional_headers(&url, &headers, &session_name, &target_service);
 
-    if req
-        .headers()
-        .get("upgrade")
-        .map(|v| v == "websocket")
-        .unwrap_or(false)
-    {
-        handle_ws_req(req, target_service, extra_headers, client).await
-    } else {
-        handle_http_req(req, target_service, extra_headers, client).await
+    match ws.0 {
+        Some(upgrade) => {
+            let mut url = target_service.url;
+            if url.starts_with("http://") {
+                url = url.replace("http://", "ws://");
+            } else if url.starts_with("https://") {
+                url = url.replace("https://", "wss://");
+            }
+
+            let uri = url.parse::<Uri>().unwrap();
+            let mut request = uri.into_client_request().unwrap();
+
+            let extra_http_headers: HeaderMap = extra_headers.into();
+            for (key, value) in extra_http_headers.iter() {
+                request.headers_mut().insert(key, value.clone());
+            }
+
+            // TODO: Handle both connection errors here and failed responses.
+            let (ws_stream, upstream_response) = tokio_tungstenite::connect_async(request)
+                .await
+                .expect("WebSocket connection failed");
```

Comment:
Match on error:

- If it's Err(HttpResponse) - there is a valid http response even though the websocket didn't connect, return the response to the client (such as 404)
- for other errors - Return a real error / 503 ish

Reply:
Changed on [0f0e525](https://github.com/mentimeter/linkup/pull/235/commits/0f0e5254b4fe14b4b4c5dbdc9601d9d2c0ea1c8c). Were you thinking of something like that?

Reply:
Super - love it